bindings/sendgrid/sendgrid.go

```diff
@@ -0,0 +1,131 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package sendgrid
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/dapr/pkg/logger"
+	"github.com/sendgrid/sendgrid-go"
+	"github.com/sendgrid/sendgrid-go/helpers/mail"
+)
+
+// SendGrid allows sending of emails using the 3rd party SendGrid service
+type SendGrid struct {
+	metadata sendGridMetadata
+	logger   logger.Logger
+}
+
+type sendGridMetadata struct {
+	APIKey    string `json:"apiKey"`
+	EmailFrom string `json:"emailFrom"`
+	EmailTo   string `json:"emailTo"`
+	Subject   string `json:"subject"`
+}
+
+// NewSendGrid returns a new SendGrid bindings instance
+func NewSendGrid(logger logger.Logger) *SendGrid {
+	return &SendGrid{logger: logger}
+}
+
+// Helper to parse metadata
+func (sg *SendGrid) parseMetadata(meta bindings.Metadata) (sendGridMetadata, error) {
+	sgMeta := sendGridMetadata{}
+
+	// Required properties
+	if val, ok := meta.Properties["apiKey"]; ok && val != "" {
+		sgMeta.APIKey = val
+	} else {
+		return sgMeta, errors.New("SendGrid binding error: apiKey field is required in metadata")
+	}
+
+	// Optional properties
+	sgMeta.EmailFrom = meta.Properties["emailFrom"]
+	sgMeta.EmailTo = meta.Properties["emailTo"]
+	sgMeta.EmailFrom = meta.Properties["emailFrom"]
+	sgMeta.Subject = meta.Properties["subject"]
+
+	return sgMeta, nil
+}
+
+// Init does metadata parsing and not much else :)
+func (sg *SendGrid) Init(metadata bindings.Metadata) error {
+	// Parse input metadata
+	meta, err := sg.parseMetadata(metadata)
+	if err != nil {
+		return err
+	}
+
+	// Um, yeah that's about it!
+	sg.metadata = meta
+	return nil
+}
+
+// Write does the work of sending message to SendGrid API
+func (sg *SendGrid) Write(req *bindings.WriteRequest) error {
+	// We allow two possible sources of these fields, the component metadata or request metadata
+	// Build email from,
+	var from *mail.Email
+	if len(sg.metadata.EmailFrom) > 0 {
+		from = mail.NewEmail("", sg.metadata.EmailFrom)
+	}
+	if len(req.Metadata["emailFrom"]) > 0 {
```

Comment:
this will overwrite the metadata.EmailFrom if set - is that the intended behaviour i.e. code then config? Feels backwards to me